internal/backend/pager/pager.go

```diff
@@ -0,0 +1,71 @@
+package pager
+
+import "fmt"
+
+const (
+	RESERVED_HEADER_PAGE_COUNT = 1
+)
+
+type file interface {
+	Read(int, int) ([]byte, error)
+	Write(int, []byte) error
+}
+
+/*
+	Pager manages the database pages for access by the BTree
+Properties
+	 - pageSize:      the number of bytes in each data page
+	 - fs:            file system operator used for reading and writing to the db
+	 - freelistIndex: the page number where the freelist starts
+
+	 The database has the following pages:
+
+	 Header:     Metadata about the database
+	 TablePages: Data pages for the database
+	 Freelist:   Unused pages to be allocated
+*/
+type Pager struct {
+	pageSize      int
+	fs            file
+	freelistIndex int
+}
+
+/*
+	Get retrieves the data page at offset i after the header
+*/
+func (p *Pager) Get(i int) ([]byte, error) {
+	if i < 0 || i >= p.freelistIndex {
+		return nil, fmt.Errorf(
+			"cannot read page %d, index must be between 0 and %d",
+			i,
+			p.freelistIndex,
+		)
+	}
+
+	offset := (i * p.pageSize) + (RESERVED_HEADER_PAGE_COUNT * p.pageSize)
+	return p.fs.Read(offset, p.pageSize)
+}
+
+/*
+	Set puts a data page at offset i with the bytes in content
+*/
+func (p *Pager) Set(i int, content []byte) error {
```

Comment:
Same here

Reply:
Also is there ever going to be a need to know where you wrote the content to? would it make sense to return the offset or something here?

Reply:
i do see the value, but eventually the pager will be behind an interface, and that interface shouldn't have a notion of offsets because its internal implementation won't matter

Reply:
oh and definitely to the descriptive name